src/Adapters/Etw.Net451/EtwTelemetryModule.cs

```diff
@@ -0,0 +1,247 @@
+                return;
+            }
+
+            if (this.isDisposed)
+            {
+                errorMessage = "Can't initialize a module that is disposed. The initialization is terminated.";
+                EventSourceListenerEventSource.Log.ModuleInitializationFailed(
+                    nameof(EtwTelemetryModule),
+                    errorMessage);
+                return;
+            }
+
+            bool? isProcessElevated = this.traceEventSession.IsElevated();
+            if (!isProcessElevated.HasValue || !isProcessElevated.Value)
+            {
+                errorMessage = "The process is required to be elevated to enable ETW providers. The initialization is terminated.";
+                EventSourceListenerEventSource.Log.ModuleInitializationFailed(
+                    nameof(EtwTelemetryModule),
+                    errorMessage);
+                
+                // Throws so that user will be able to see the exception message in Output Window for debugging.
+                throw new UnauthorizedAccessException(errorMessage);
+            }
+
+            lock (this.lockObject)
+            {
+                this.client = new TelemetryClient(configuration);
+
+                // sdkVersionIdentifier will be used in telemtry entry as a identifier for the sender.
+                // The value will look like: etw:x.x.x-x
+                const string SdkVersionIdentifier = "etw:";
+                this.client.Context.GetInternalContext().SdkVersion = SdkVersionUtils.GetSdkVersion(SdkVersionIdentifier);
+
+                if (this.isInitialized)
+                {
+                    this.isInitialized = false;
+                    this.DisableProviders();
+                    this.enabledProviderIds.Clear();
+                    this.enabledProviderNames.Clear();
+                }
+
+                if (this.Sources.Count == 0)
+                {
+                    EventSourceListenerEventSource.Log.NoSourcesConfigured(moduleName: nameof(EtwTelemetryModule));
+                    return;
+                }
+
+                this.EnableProviders();
+                try
+                {
+                    // Start the trace session
+                    Task.Factory.StartNew(() => this.StartTraceEventSession(this.traceEventSession, this.client), TaskCreationOptions.LongRunning);
+                }
+                finally
+                {
+                    this.isInitialized = true;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Disposes the module.
+        /// </summary>
+        public void Dispose()
+        {
+            this.Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        /// <summary>
+        /// Disposes the module.
+        /// </summary>
+        /// <param name="isDisposing">Indicate if it is called by Dispose().</param>
+        protected virtual void Dispose(bool isDisposing)
+        {
+            if (this.isDisposed)
```

Comment:
Sorry to make you state your point again because I've been mix things for my first comment. Let me try again to see if I can make it clear.

1. Your major concern is that it is not clear if Dispose is allowed to be called more than once, right?
And the answer is yes, it is safe here when Dispose(bool) is called more than once.

2. Your second concern is that the check for 'isDisposed' is not enough a guarantee of code being called more than once.
And that is correct. Since Dispose() is allowed to be called more than once concurrently, it is OK if more than one threads pass the point of the check.
One of the small benefit of having the check is outside the scenario of concurrency calling, it will save a little bit calling into traceEventSession.Dispose(). And that is it.

3. Let me know if bullet 1 or 2 is not clear or if you still think I missed some points because I want to full understand your idea before taking actions on the code.

Also let me know if you still feel uncomfortable of having the check of 'isDisposed' and I'll take it as a final decision and I'll remove it from the code.